### PR TITLE
Fix password complexity regex for special characters (on master)

### LINKED ISSUE #8521: account password recovery error 500
- Gitea version (or commit ref): 1.10.0-rc1
- Git version: 2.17.1
- Operating system: Ubuntu 18.04.3 LTS
- Database (use `[x]`):
  - [x] MySQL
- Can you reproduce the bug at https://try.gitea.io:
  - [x] Yes

Upgrading from 1.9.4 to 1.10.0-rc1 breaks account password recovery after I enter the new email address and click the "Recover Account" button. There's the call stack.

2019/10/15 06:07:56 ...les/context/panic.go:35:1() [E] PANIC:: regexp: Compile("][ !\"#$%&'()*+,./:;<=>?@\\\\^_{|}~`-"): error parsing regexp: missing closing ]: `[ !"#$%&'()*+,./:;<=>?@\\^_{|}~`-`
        /usr/local/go/src/regexp/regexp.go:311 (0x6bdaf1)
        /go/src/code.gitea.io/gitea/modules/password/password.go:31 (0x10a2214)
        /usr/local/go/src/sync/once.go:66 (0x484b02)
        /usr/local/go/src/sync/once.go:57 (0x10a1c3c)
        /go/src/code.gitea.io/gitea/modules/password/password.go:28 (0x10a1c0d)
        /go/src/code.gitea.io/gitea/modules/password/password.go:46 (0x10a1ca2)
        /go/src/code.gitea.io/gitea/routers/user/auth.go:1338 (0x131e000)
        /usr/local/go/src/reflect/value.go:460 (0x4977d5)
        /usr/local/go/src/reflect/value.go:321 (0x496f93)
        /go/src/code.gitea.io/gitea/vendor/gitea.com/macaron/inject/inject.go:177 (0x9a6bf9)
        /go/src/code.gitea.io/gitea/vendor/gitea.com/macaron/inject/inject.go:137 (0x9a65a9)
        /go/src/code.gitea.io/gitea/vendor/gitea.com/macaron/macaron/context.go:121 (0x9d5f38)
        /go/src/code.gitea.io/gitea/vendor/gitea.com/macaron/macaron/context.go:112 (0x112eb11)
        /go/src/code.gitea.io/gitea/modules/context/panic.go:39 (0x112eafd)
        /usr/local/go/src/reflect/value.go:460 (0x4977d5)
        /usr/local/go/src/reflect/value.go:321 (0x496f93)
        /go/src/code.gitea.io/gitea/vendor/gitea.com/macaron/inject/inject.go:177 (0x9a6bf9)
        /go/src/code.gitea.io/gitea/vendor/gitea.com/macaron/inject/inject.go:137 (0x9a65a9)
        /go/src/code.gitea.io/gitea/vendor/gitea.com/macaron/macaron/context.go:121 (0x9d5f38)
        /go/src/code.gitea.io/gitea/vendor/gitea.com/macaron/macaron/context.go:112 (0xaa35f5)
        /go/src/code.gitea.io/gitea/vendor/gitea.com/macaron/session/session.go:192 (0xaa35e0)
        /go/src/code.gitea.io/gitea/vendor/gitea.com/macaron/macaron/context.go:79 (0x9d5de0)
        /go/src/code.gitea.io/gitea/vendor/gitea.com/macaron/inject/inject.go:157 (0x9a6909)
        /go/src/code.gitea.io/gitea/vendor/gitea.com/macaron/inject/inject.go:135 (0x9a6698)
        /go/src/code.gitea.io/gitea/vendor/gitea.com/macaron/macaron/context.go:121 (0x9d5f38)
        /go/src/code.gitea.io/gitea/vendor/gitea.com/macaron/macaron/context.go:112 (0x9e75d9)
        /go/src/code.gitea.io/gitea/vendor/gitea.com/macaron/macaron/recovery.go:161 (0x9e75c7)
        /go/src/code.gitea.io/gitea/vendor/gitea.com/macaron/macaron/logger.go:40 (0x9d9ab3)
        /go/src/code.gitea.io/gitea/vendor/gitea.com/macaron/inject/inject.go:157 (0x9a6909)
        /go/src/code.gitea.io/gitea/vendor/gitea.com/macaron/inject/inject.go:135 (0x9a6698)
        /go/src/code.gitea.io/gitea/vendor/gitea.com/macaron/macaron/context.go:121 (0x9d5f38)
        /go/src/code.gitea.io/gitea/vendor/gitea.com/macaron/macaron/context.go:112 (0x9e6910)
        /go/src/code.gitea.io/gitea/vendor/gitea.com/macaron/macaron/logger.go:52 (0x9e68fb)
        /go/src/code.gitea.io/gitea/vendor/gitea.com/macaron/macaron/logger.go:40 (0x9d9ab3)
        /go/src/code.gitea.io/gitea/vendor/gitea.com/macaron/inject/inject.go:157 (0x9a6909)
        /go/src/code.gitea.io/gitea/vendor/gitea.com/macaron/inject/inject.go:135 (0x9a6698)
        /go/src/code.gitea.io/gitea/vendor/gitea.com/macaron/macaron/context.go:121 (0x9d5f38)
        /go/src/code.gitea.io/gitea/vendor/gitea.com/macaron/macaron/router.go:187 (0x9e87f6)
        /go/src/code.gitea.io/gitea/vendor/gitea.com/macaron/macaron/router.go:294 (0x9e227d)
        /go/src/code.gitea.io/gitea/vendor/gitea.com/macaron/macaron/macaron.go:220 (0x9dae4a)
        /go/src/code.gitea.io/gitea/vendor/github.com/gorilla/context/context.go:141 (0xc4104d)
        /usr/local/go/src/net/http/server.go:2007 (0x754133)
        /usr/local/go/src/net/http/server.go:2802 (0x757583)
        /usr/local/go/src/net/http/server.go:1890 (0x752f24)
        /usr/local/go/src/runtime/asm_amd64.s:1357 (0x466bd0)

### PR DESCRIPTION
**Pull request by guillep2k**
Fixes #8521

### CODE REVIEW

**Comment by guillep2k** [comment]
Also fixes the error display in the change password page.

**Review by zeripath — APPROVED** [review]

File: modules/setting/setting.go [hunk 1]
```diff
@@ -785,7 +785,7 @@ func NewContext() {
 		"lower": "[a-z]+",
 		"upper": "[A-Z]+",
 		"digit": "[0-9]+",
-		"spec":  `][ !"#$%&'()*+,./:;<=>?@\\^_{|}~` + "`-",
+		"spec":  `[][ !"#$%&'()*+,./:;<=>?@\\^_{|}~` + "`-]",
```
Comment (lunny) [inline modules/setting/setting.go]: Maybe you should add a `+` at the end of the string.

**Review by lunny — CHANGES_REQUESTED** [review]

**Comment by lunny** [comment]
To avoid the bug again, please move `dictPC` as a gloabl variable and add a test for it.

**Comment by guillep2k** [comment]
> 
> 
> To avoid the bug again, please move `dictPC` as a gloabl variable and add a test for it.

Will do

**Comment by guillep2k** [comment]
@lunny @zeripath Done.

* Removed regexp completely as they complicated things unnecessarily.
* Added an 'off' setting as there was no way to disable the functionality.
* Established an `[a-zA-Z0-9]` default for password generation pattern for when complexity checks are off.
* Fixed cheat sheet and ini sample accordingly.
* Added tests.

**Review by zeripath — APPROVED** [review]

**Review by adelowo — APPROVED** [review]

File: modules/password/password.go [hunk 2]
```diff
@@ -7,45 +7,58 @@ package password
...
 	"math/big"
-	"regexp"
+	"strings"
 	"sync"
 
 	"code.gitea.io/gitea/modules/setting"
 )
 
-var matchComplexities = map[string]regexp.Regexp{}
 var matchComplexityOnce sync.Once
```
Comment (lunny) [inline modules/password/password.go]: A nit thing, group all the variables will look better.
```
var (
  matchComplexityOnce sync.Once
  validChars string
)
```

File: modules/password/password.go [hunk 3]
```diff
@@ -7,45 +7,58 @@ package password
...
 
-// IsComplexEnough return True if password is Complexity
+// IsComplexEnough return True if password meets complexity settings
 func IsComplexEnough(pwd string) bool {
-	if len(setting.PasswordComplexity) > 0 {
-		NewComplexity()
-		for _, val := range matchComplexities {
-			if !val.MatchString(pwd) {
+	NewComplexity()
+	if len(validChars) > 0 {
```
Comment (lunny) [inline modules/password/password.go:61]: It seems `validChars` only is used to know if we need check the complexity? How about remove `validChars` and only use `requiredChars`.

Comment (guillep2k) [reply modules/password/password.go:61]: `validChars` is used in `Generate()` to have the list of characters to choose from.
`requiredChars` is used in `IsComplexEnough()` to check the password against.

They require almost the same information, except that `Generate()` benefits from having all the characters in one string (and always requires this input) while `IsComplexEnough()` will cycle through all groups checking that the password contains at least one character from each group.

File: modules/password/password.go (hunk 2 above)
Comment (guillep2k) [reply modules/password/password.go]: @lunny done.

Comment (lunny) [reply modules/password/password.go]: It seems not?

Comment (guillep2k) [reply modules/password/password.go]: @lunny Sorry, I had pushed to the backport branch. Now I pushed to this branch.

**Review by lunny — APPROVED** [review]

**Pull request merged**